### PR TITLE
[Fabric] Shim RCTTextInput

### LINKED ISSUE #1547: Shim RCTTextInput

### PR DESCRIPTION
**Pull request by shwanton**
#### Please select one of the following
- [x] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [ ] I am making a fix / change for the macOS implementation of react-native
- [ ] I am making a change required for Microsoft usage of react-native

## Summary

Fabric on macOS implementation:
Shim RCTTextInput to work w/ Fabric

closes #1547 

## Changelog

[macOS][Added] - Shim RCTTextInput for Fabric

## Test Plan

[x] Build RNTester-macOS w/ Fabric - doesn’t run yet, but no RCTTextInput errors
CleanShot 2023-01-23 at 16 51 43

[x] Build RNTester - iOS w/ Fabric
CleanShot 2023-01-23 at 16 53 04

[x] Build RNTester-macOS w/ Paper - should work
CleanShot 2023-01-23 at 16 59 48

[x] Build RNTester - iOS w/ Paper -
CleanShot 2023-01-23 at 16 58 23

### CODE REVIEW

File: React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm [hunk 1]
```diff
@@ -30,7 +30,11 @@ @interface RCTTextInputComponentView () <RCTBackedTextInputDelegate, RCTTextInpu
 
 @implementation RCTTextInputComponentView {
   TextInputShadowNode::ConcreteState::Shared _state;
-  RCTUIView<RCTBackedTextInputViewProtocol> *_backedTextInputView; // [macOS]
+#if !TARGET_OS_OSX // [macOS]
+  RCTUIView<RCTBackedTextInputViewProtocol> *_backedTextInputView;
+#else // [macOS
+  RCTUITextView<RCTBackedTextInputViewProtocol> *_backedTextInputView;
+#endif // macOS]
```
Comment (Saadnajmi) [inline React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:37]: This feels like a significant change from Core?

File: React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm [hunk 2]
```diff
@@ -478,25 +486,23 @@ - (void)setTextAndSelection:(NSInteger)eventCount
...
 
   // These keyboard types (all are number pads) don't have a "Done" button by default,
   // so we create an `inputAccessoryView` with this button for them.
   BOOL shouldHaveInputAccesoryView =
-      (keyboardType == UIKeyboardTypeNumberPad || keyboardType == UIKeyboardTypePhonePad ||
-       keyboardType == UIKeyboardTypeDecimalPad || keyboardType == UIKeyboardTypeASCIICapableNumberPad) &&
-      _backedTextInputView.returnKeyType == UIReturnKeyDone;
+  (keyboardType == UIKeyboardTypeNumberPad || keyboardType == UIKeyboardTypePhonePad ||
+   keyboardType == UIKeyboardTypeDecimalPad || keyboardType == UIKeyboardTypeASCIICapableNumberPad) &&
+  _backedTextInputView.returnKeyType == UIReturnKeyDone;
```
Comment (Saadnajmi) [inline React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:505]: Is this coming from a formatter? What does upstream have?

**Review by Saadnajmi — COMMENTED** [review]
Note: This might change a bit with 0.71 because 0.71 introduced this commit (https://github.com/facebook/react-native/pull/33653) that conflicts with some of the macOS specific changes we made to TextInput (https://github.com/microsoft/react-native-macos/pull/1423). So this PR might be worth holding off till after 0.71.

File: React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm (hunk 1 above)
Comment (shwanton) [reply React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:37]: We plan to spend a good chunk of time on TextInput due to the complexity. If we can shim this for the short-term & then revisit once 0.71 lands and we properly adapt for Fabric.

File: React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm (hunk 2 above)
Comment (shwanton) [reply React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:505]: This spacing is from core:
https://github.com/microsoft/react-native-macos/blob/main/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L496-L499

**Comment by shwanton** [comment]
> Note: This might change a bit with 0.71 because 0.71 introduced this commit ([facebook#33653](https://github.com/facebook/react-native/pull/33653)) that conflicts with some of the macOS specific changes we made to TextInput (#1423). So this PR might be worth holding off till after 0.71.

Once we have Fabric building, we plan to revisit TextInput in depth & use the latest Fabric changes.

**Review by Saadnajmi — APPROVED** [review]

**Pull request merged**